src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala

```diff
@@ -0,0 +1,47 @@
+package com.ambiata.poacher.hdfs
+
+import com.ambiata.mundane.control._
+import com.ambiata.mundane.io._
+import com.ambiata.mundane.io.Temporary._
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import java.util.concurrent.atomic.AtomicInteger
+import scalaz._, Scalaz._
+
+case class HdfsTemporary(seed: String) {
+  private val step: AtomicInteger = new AtomicInteger(0)
+
+  def path: Hdfs[Path] = {
+    val base = uniqueDirPath
+    val incr = step.incrementAndGet.toString
+    val path = new Path(new Path(base.path, incr), seed)
+    val msg = s"Hdfs($path)"
+    addCleanupFinalizer(new Path(base.path), msg) >>
+      addPrintFinalizer(msg) >>
+        Hdfs.ok(path)
+  }
+
+  def addCleanupFinalizer(path: Path, msg: String): Hdfs[Unit] =
+    if (skipCleanup) forceAddPrintFinalizer(msg)
+    else             Hdfs.addFinalizer(Hdfs.deleteAll(path))
+
+  def addPrintFinalizer(msg: String): Hdfs[Unit] =
+    if (print && !skipCleanup) forceAddPrintFinalizer(msg)
+    else                       Hdfs.unit
+
+  def forceAddPrintFinalizer(msg: String): Hdfs[Unit] =
+    Hdfs.addFinalizer(Hdfs.putStrLn(s"Temporary: $msg"))
+}
+
+object HdfsTemporary {
+  def getTemporaryHdfsX: Hdfs[Path] = {
+    val p = new Path(uniqueDirPath.path)
+    addHdfsFinalizer(p) >> Hdfs.ok(p)
```

Comment:
> Add a finalizer for p, so p gets cleaned up and return p. P being a Path

`as` is a extension method for `Functor` that's the same as doing `something.map(_ => x)`, which is what you're doing above. :)